custom_components/amplipi/media_player.py

```diff
@@ -1235,3 +1240,288 @@ async def async_set_volume_level(self, volume):
+            await self.async_connect_stream_to_source(self._stream)
+        elif self._stream.type == "rca": # Autoconnect stream if it can only ever connect to one source, regardless of space
+            goal_source = next((s for s in self._sources if s.id == self._id - 996), None)
+            if goal_source:
+                await self.async_connect_stream_to_source(self._stream, goal_source)
+
+        self._is_off = False
+
+    async def async_turn_off(self):
+        try:
+            if self._current_source is not None:
+                _LOGGER.info(f"Disconnecting stream from source {self._current_source.name}")
+                await self._update_zones(
+                    ZoneUpdate(
+                        source_id=-1,
+                    )
+                )
+                await self._client.set_source(
+                    self._current_source.id,
+                    SourceUpdate(
+                        input='None'
+                    )
+                )
+        except AttributeError:
+            # There can be a race condition where self._current_source is not None and then becomes None by the time self._current_source.name or self._current_source.id is read
+            # This happens if the user (or another entity/automation) disconnects the stream from the source while (or shortly before, due to the polling rate) this function runs
+            # I'd rather suppress the error and ensure that the current state is as accurate as possible due to the error functionally being "Cannot disconnect due to already being disconnected"
+            _LOGGER.debug(f"{self._name} had trouble disconnecting from a source")
+        finally:
+            self._is_off = True
+            
+
+    async def async_mute_volume(self, mute):
+        if mute is None:
+            return
+
+        if self._current_source is not None:
+            _LOGGER.info(f"setting mute to {mute}")
+            await self._update_zones(
+                ZoneUpdate(
+                    mute=mute,
+                )
+            )
+
+    async def async_set_volume_level(self, volume):
+        if volume is None:
+            return
+        if self._current_source is not None:
+            await self._update_zones(
+                ZoneUpdate(
+                    vol_f=volume,
+                    mute=False,
+                )
+            )
+
+    @property
+    def supported_features(self):
+        """Return flag of media commands that are supported."""
+
+        supported_features = SUPPORT_AMPLIPI_DAC
+        if self._current_source is not None and self._current_source.info is not None and len(self._current_source.info.supported_cmds) > 0:
+            supported_features = supported_features | reduce(
+                operator.or_,
+                [
+                    SUPPORT_LOOKUP_DICT.get(key) for key
+                    in (SUPPORT_LOOKUP_DICT.keys() & self._current_source.info.supported_cmds)
+                ]
+            )
```

Comment:
I think this could be a little more readable, here's my first stab:

```suggestion
        if self._current_source is not None and self._current_source.info is not None:
            cmds_avail = SUPPORT_LOOKUP_DICT.keys() & self._current_source.info.supported_cmds
            features_avail = [SUPPORT_LOOKUP_DICT.get(cmd) for cmd in cmds_avail]
            supported_features |= reduce(operator.or_,  features_avail)
```